File: src/main/java/pl/mjedynak/idea/plugins/builder/verifier/PsiFieldVerifier.java
package pl.mjedynak.idea.plugins.builder.verifier;

import static org.apache.commons.lang3.StringUtils.EMPTY;

import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiField;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiModifier;
import com.intellij.psi.PsiModifierList;
import com.intellij.psi.PsiParameter;
import com.intellij.psi.PsiParameterList;
import org.apache.commons.text.WordUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.VisibleForTesting;
import pl.mjedynak.idea.plugins.builder.settings.CodeStyleSettings;

public class PsiFieldVerifier {

    @VisibleForTesting
    static final String SET_PREFIX = "set";

    @VisibleForTesting
    static final String GET_PREFIX = "get";

    private PsiFieldVerifier() {
        throw new UnsupportedOperationException("Utility class");
    }

    public static boolean isSetInConstructor(@NotNull PsiField psiField, @NotNull PsiClass psiClass) {
        boolean result = false;
        PsiMethod[] constructors = psiClass.getConstructors();
        for (int i = 0; i < constructors.length && !result; i++) {
            result = checkConstructor(psiField, constructors[i]);
        }
        return result;
    }

    public static boolean checkConstructor(@NotNull PsiField psiField, @NotNull PsiMethod constructor) {
        PsiParameterList parameterList = constructor.getParameterList();
        PsiParameter[] parameters = parameterList.getParameters();
        return iterateOverParameters(psiField, parameters);
    }

    private static boolean iterateOverParameters(
            @NotNull PsiField psiField, @NotNull PsiParameter @NotNull [] parameters) {
        boolean result = false;
        for (int i = 0; i < parameters.length && !result; i++) {
            result = areNameAndTypeEqual(psiField, parameters[i]);
        }
        return result;
    }

    public static boolean areNameAndTypeEqual(@NotNull PsiField psiField, @NotNull PsiParameter parameter) {
        String parameterNamePrefix = CodeStyleSettings.PARAMETER_NAME_PREFIX;
        String parameterName = parameter.getName();
        String parameterNameWithoutPrefix = parameterName.replace(parameterNamePrefix, "");
        String fieldNamePrefix = CodeStyleSettings.FIELD_NAME_PREFIX;
        String fieldName = psiField.getName();
        String fieldNameWithoutPrefix = fieldName.replaceFirst(fieldNamePrefix, "");
        return parameterNameWithoutPrefix.equals(fieldNameWithoutPrefix)
                && parameter.getType().equals(psiField.getType());
    }

    public static boolean isSetInSetterMethod(@NotNull PsiField psiField, @NotNull PsiClass psiClass) {
        return methodIsNotPrivateAndHasProperPrefixAndProperName(psiField, psiClass, SET_PREFIX);
    }

    public static boolean hasGetterMethod(@NotNull PsiField psiField, @NotNull PsiClass psiClass) {
        return methodIsNotPrivateAndHasProperPrefixAndProperName(psiField, psiClass, GET_PREFIX);
    }

    private static boolean methodIsNotPrivateAndHasProperPrefixAndProperName(
            @NotNull PsiField psiField, @NotNull PsiClass psiClass, @NotNull String prefix) {
        boolean result = false;
        for (PsiMethod method : psiClass.getAllMethods()) {
            if (methodIsNotPrivate(method) && methodHaProperPrefixAndProperName(psiField, method, prefix)) {
                result = true;
                break;
            }
        }
        return result;
    }

    private static boolean methodIsNotPrivate(@NotNull PsiMethod method) {
        PsiModifierList modifierList = method.getModifierList();
        return modifierListHasNoPrivateModifier(modifierList);
    }

    private static boolean methodHaProperPrefixAndProperName(
            @NotNull PsiField psiField, @NotNull PsiMethod method, @NotNull String prefix) {
        String fieldNamePrefix = CodeStyleSettings.FIELD_NAME_PREFIX;
        String fieldNameWithoutPrefix = psiField.getName().replace(fieldNamePrefix, EMPTY);
        return method.getName().equals(prefix + WordUtils.capitalize(fieldNameWithoutPrefix));
    }

    private static boolean modifierListHasNoPrivateModifier(@NotNull PsiModifierList modifierList) {
        return !modifierList.hasExplicitModifier(PsiModifier.PRIVATE);
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/verifier/PsiFieldVerifier.kt
package io.github.kingg22.kobra.builder.verifier

import com.intellij.psi.PsiClass
import com.intellij.psi.PsiField
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiModifier
import com.intellij.psi.PsiModifierList
import com.intellij.psi.PsiParameter
import io.github.kingg22.kobra.builder.settings.CodeStyleSettings
import org.apache.commons.text.WordUtils
import org.jetbrains.annotations.VisibleForTesting

public object PsiFieldVerifier {
    @VisibleForTesting
    public const val SET_PREFIX: String = "set"

    @VisibleForTesting
    public const val GET_PREFIX: String = "get"

    @JvmStatic
    public fun isSetInConstructor(psiField: PsiField, psiClass: PsiClass): Boolean {
        var result = false
        val constructors = psiClass.constructors
        var i = 0
        while (i < constructors.size && !result) {
            result = checkConstructor(psiField, constructors[i]!!)
            i++
        }
        return result
    }

    @JvmStatic
    public fun checkConstructor(psiField: PsiField, constructor: PsiMethod): Boolean {
        val parameterList = constructor.parameterList
        val parameters = parameterList.parameters
        return iterateOverParameters(psiField, parameters)
    }

    private fun iterateOverParameters(psiField: PsiField, parameters: Array<PsiParameter>): Boolean {
        var result = false
        var i = 0
        while (i < parameters.size && !result) {
            result = areNameAndTypeEqual(psiField, parameters[i])
            i++
        }
        return result
    }

    public fun areNameAndTypeEqual(psiField: PsiField, parameter: PsiParameter): Boolean {
        val parameterNamePrefix: String = CodeStyleSettings.PARAMETER_NAME_PREFIX
        val parameterName = parameter.name
        val parameterNameWithoutPrefix = parameterName.replace(parameterNamePrefix, "")
        val fieldNamePrefix: String = CodeStyleSettings.FIELD_NAME_PREFIX
        val fieldName = psiField.name
        val fieldNameWithoutPrefix = fieldName.replaceFirst(fieldNamePrefix.toRegex(), "")
        return parameterNameWithoutPrefix == fieldNameWithoutPrefix &&
            parameter.type == psiField.type
    }

    @JvmStatic
    public fun isSetInSetterMethod(psiField: PsiField, psiClass: PsiClass): Boolean =
        methodIsNotPrivateAndHasProperPrefixAndProperName(psiField, psiClass, SET_PREFIX)

    @JvmStatic
    public fun hasGetterMethod(psiField: PsiField, psiClass: PsiClass): Boolean =
        methodIsNotPrivateAndHasProperPrefixAndProperName(psiField, psiClass, GET_PREFIX)

    private fun methodIsNotPrivateAndHasProperPrefixAndProperName(
        psiField: PsiField,
        psiClass: PsiClass,
        prefix: String,
    ): Boolean {
        var result = false
        for (method in psiClass.allMethods) {
            if (methodIsNotPrivate(method) && methodHaProperPrefixAndProperName(psiField, method, prefix)) {
                result = true
                break
            }
        }
        return result
    }

    private fun methodIsNotPrivate(method: PsiMethod): Boolean {
        val modifierList = method.modifierList
        return modifierListHasNoPrivateModifier(modifierList)
    }

    private fun methodHaProperPrefixAndProperName(psiField: PsiField, method: PsiMethod, prefix: String): Boolean {
        val fieldNamePrefix = CodeStyleSettings.FIELD_NAME_PREFIX
        val fieldNameWithoutPrefix = psiField.name.replace(fieldNamePrefix, "")
        return method.name == prefix + WordUtils.capitalize(fieldNameWithoutPrefix)
    }

    private fun modifierListHasNoPrivateModifier(modifierList: PsiModifierList): Boolean =
        !modifierList.hasExplicitModifier(PsiModifier.PRIVATE)
}
